services/src/main/java/org/keycloak/storage/GroupStorageManager.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.storage;
+
+import org.keycloak.component.ComponentModel;
+import org.keycloak.models.GroupModel;
+import org.keycloak.models.GroupProvider;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.RoleModel;
+import org.keycloak.storage.group.GroupLookupProvider;
+import org.keycloak.storage.group.GroupStorageProvider;
+import org.keycloak.storage.group.GroupStorageProviderModel;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class GroupStorageManager extends AbstractStorageManager<GroupStorageProvider, GroupStorageProviderModel> implements GroupProvider {
+
+    public GroupStorageManager(KeycloakSession session) {
+        super(session, GroupStorageProvider.class, GroupStorageProviderModel::new, "group");
+    }
+
+    /* GROUP PROVIDER LOOKUP METHODS - implemented by group storage providers */
+
+    @Override
+    public GroupModel getGroupById(RealmModel realm, String id) {
+        StorageId storageId = new StorageId(id);
+        if (storageId.getProviderId() == null) {
+            return session.groupLocalStorage().getGroupById(realm, id);
+        }
+
+        GroupLookupProvider provider = getStorageProviderInstance(realm, storageId.getProviderId());
+        if (provider == null) return null;
+
+        return provider.getGroupById(realm, id);
+    }
+
+    /**
+     * Obtaining groups from an external client storage is time-bounded. In case the external group storage
+     * isn't available at least groups from a local storage are returned. For this purpose
+     * the {@link org.keycloak.services.DefaultKeycloakSessionFactory#getClientStorageProviderTimeout()} property is used.
+     * Default value is 3000 milliseconds and it's configurable.
+     * See {@link org.keycloak.services.DefaultKeycloakSessionFactory} for details.
+     *
+     * TODO: Remove stream() and .collect(Collectors.toList()) during streamification
```

Comment:
Is this still valid?

Reply:
I think so. This PR: https://github.com/keycloak/keycloak/pull/7328 should be merged first and then I will resolve this TODO

Reply:
oh I see, I thought this is another way around, thanks

Reply:
#7328 has been merged just now.